genkit-tools/common/src/eval/localFileDatasetStore.ts

```diff
@@ -0,0 +1,221 @@
+    if (!fs.existsSync(this.indexFile)) {
+      fs.writeFileSync(path.resolve(this.indexFile), '');
+    }
+    logger.info(
+      `Initialized local file dataset store at root: ${this.storeRoot}`
+    );
+  }
+
+  static getDatasetStore() {
+    if (!this.cachedDatasetStore) {
+      this.cachedDatasetStore = new LocalFileDatasetStore();
+    }
+    return this.cachedDatasetStore;
+  }
+
+  static reset() {
+    this.cachedDatasetStore = null;
+  }
+
+  async createDataset(req: CreateDatasetRequest): Promise<DatasetMetadata> {
+    return this.createDatasetInternal(req.data, req.displayName);
+  }
+
+  private async createDatasetInternal(
+    data: Dataset,
+    displayName?: string
+  ): Promise<DatasetMetadata> {
+    const datasetId = this.generateDatasetId();
+    const filePath = path.resolve(
+      this.storeRoot,
+      this.generateFileName(datasetId)
+    );
+
+    if (!fs.existsSync(filePath)) {
+      logger.error(`Dataset already exists at ` + filePath);
+      throw new Error(
+        `Create dataset failed: file already exists at {$filePath}`
+      );
+    }
+
+    logger.info(`Saving Dataset to ` + filePath);
+    await writeFile(filePath, JSON.stringify(data));
+
+    const metadata = {
+      datasetId: datasetId,
+      size: Array.isArray(data) ? data.length : data.samples.length,
+      uri: 'file://' + filePath,
+      version: 1,
+      displayName: displayName,
+      createTime: Date.now().toString(),
+      updateTime: Date.now().toString(),
+    };
+
+    logger.debug(
+      `Saving DatasetMetadata for ID ${datasetId} to ` +
+        path.resolve(this.indexFile)
+    );
+    await appendFile(
+      path.resolve(this.indexFile),
+      JSON.stringify(metadata) + this.INDEX_DELIMITER
+    );
+    return metadata;
+  }
+
+  async updateDataset(req: UpdateDatasetRequest): Promise<DatasetMetadata> {
+    if (!req.datasetId) {
+      return this.createDatasetInternal(req.patch, req.displayName);
+    }
+    const datasetId = req.datasetId;
+    const filePath = path.resolve(
+      this.storeRoot,
+      this.generateFileName(datasetId)
+    );
+    if (!fs.existsSync(filePath)) {
+      return this.createDatasetInternal(req.patch, req.displayName);
+    }
+
+    const metadatas = await this.listDatasets();
+    const prevMetadata = metadatas.find(
+      (metadata) => metadata.datasetId === datasetId
+    );
+    if (!prevMetadata) {
+      throw new Error(`Update dataset failed: dataset metadata not found`);
+    }
+
+    logger.info(`Updating Dataset at ` + filePath);
+    await writeFile(filePath, JSON.stringify(req.patch));
+
+    const newMetadata = {
+      datasetId: datasetId,
+      size: Array.isArray(req.patch)
+        ? req.patch.length
+        : req.patch.samples.length,
+      uri: 'file://' + filePath,
+      version: prevMetadata.version++,
+      displayName: req.displayName,
+      createTime: prevMetadata.createTime,
+      updateTime: Date.now().toString(),
+    };
```

Comment:
This looks like it is overriding _all_ rows during update. This works for now, but it had a more efficient upsert implementation in mind (add new rows, and update existing rows, identify rows by deterministicHash(row)). WDYT?

Reply:
This also means that we ill have to slightly change the way we store the dataset files (keep a row identifier on file)

Reply:
Done.